homeassistant/components/matrix.py

```diff
@@ -0,0 +1,358 @@
+            cmd = pieces[0][1:]
+
+            command = self._word_commands.get(room_id, {}).get(cmd)
+            if command:
+                event_data = {
+                    'command': command[CONF_NAME],
+                    'sender': event['sender'],
+                    'room': room_id,
+                    'args': pieces[1:]
+                }
+                self.hass.bus.fire(EVENT_MATRIX_COMMAND, event_data)
+
+        # After single-word commands, check all regex commands in the room
+        for command in self._expression_commands.get(room_id, []):
+            match = command[CONF_EXPRESSION].match(event['content']['body'])
+            if not match:
+                continue
+            event_data = {
+                'command': command[CONF_NAME],
+                'sender': event['sender'],
+                'room': room_id,
+                'args': match.groupdict()
+            }
+            self.hass.bus.fire(EVENT_MATRIX_COMMAND, event_data)
+
+    def _join_or_get_room(self, room_id_or_alias):
+        """Join a room or get it, if we are already in the room.
+
+        We can't just always call join_room(), since that seems to crash
+        the client if we're already in the room.
+        """
+        rooms = self._client.get_rooms()
+        if room_id_or_alias in rooms:
+            _LOGGER.debug("Already in room %s", room_id_or_alias)
+            return rooms[room_id_or_alias]
+
+        for room in rooms.values():
+            if room.room_id not in self._aliases_fetched_for:
+                room.update_aliases()
+                self._aliases_fetched_for.add(room.room_id)
+
+            if room_id_or_alias in room.aliases:
+                _LOGGER.debug("Already in room %s (known as %s)",
+                              room.room_id, room_id_or_alias)
+                return room
+
+        room = self._client.join_room(room_id_or_alias)
+        _LOGGER.info("Joined room %s (known as %s)", room.room_id,
+                     room_id_or_alias)
+        return room
+
+    def _join_rooms(self):
+        """Join the rooms that we listen for commands in."""
+        from matrix_client.client import MatrixRequestError
+
+        for room_id in self._listening_rooms:
+            try:
+                room = self._join_or_get_room(room_id)
+                room.add_listener(partial(self._handle_room_message, room_id),
+                                  "m.room.message")
+
+            except MatrixRequestError as ex:
+                _LOGGER.error("Could not join room %s: %s", room_id, ex)
+
+    def _get_auth_tokens(self):
+        """
+        Read sorted authentication tokens from disk.
+
+        Returns the auth_tokens dictionary.
+        """
+        if not os.path.exists(self._session_filepath):
+            return {}
+
+        try:
+            data = load_json(self._session_filepath)
+
+            auth_tokens = {}
+            for mx_id, token in data.items():
```

Comment:
I'm confused, isn't `data` already the dict that you need? And if so, couldn't you just use `load_json` directly, since  it will also handle file not found.

Reply:
It is… I think I took that code from the old matrix notifier platform without too much thought… Sorry.